### ISSUE TITLE
Very slow performance when reproducing DFlash on Qwen3-VL 32B

### ISSUE THREAD

**Issue by Liac-li** [z-lab/dflash#79]
I am experiencing severe performance degradation when using **DFlash** speculative decoding with **Qwen3-VL 32B** in `sglang`. 

We trained a custom DFlash drafter for Qwen3-VL 32B (using ~12K data including Caption, GUI grounding, and Math reasoning for 2 epochs, evaluated on H100 GPUs，5 draft layers ) with `SpecForge`. While we expected speculative decoding to accelerate inference, testing with the `sglang benchserver` on the MMMU dataset reveals a massive slowdown and extreme latency overheads compared to the standard backend.

### Performance Breakdown (MMMU Benchmark)

Here is the direct comparison of the key metrics (Vanilla vs. DFlash):

| Metric | Vanilla Qwen3-VL-32B | Qwen3-VL-32B + DFlash |
| :--- | :--- | :--- |
| **Output Token Throughput** | 6432.85 tok/s | **958.33 tok/s** |
| **Mean TTFT** | 24.6 seconds | **238.8 seconds** |
| **Mean TPOT** | 26.14 ms | **48.34 ms** |
| **Benchmark Duration** | 80.39 s | **539.61 s** |
| **Accept Length** | N/A | **1.24** |  |

### Raw Benchmark Logs

1. Qwen3-VL-32B (Vanilla Baseline on MMMU)</b>
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 512       
Successful requests:                     505       
Benchmark duration (s):                  80.39     
Total input tokens:                      169295    
Total input text tokens:                 38928     
Total input vision tokens:               130367    
Total generated tokens:                  517120    
Total generated tokens (retokenized):    486851    
Request throughput (req/s):              6.28      
Input token throughput (tok/s):          2105.99   
Output token throughput (tok/s):         6432.85   
Peak output token throughput (tok/s):    9019.00   
...
Mean E2E Latency (ms):                   51375.00  
Mean TTFT (ms):                          24635.29  
Mean TPOT (ms):                          26.14     
==================================================
```
2. Qwen3-VL-32B + DFlash (on MMMU)
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 512       
Successful requests:                     505       
Benchmark duration (s):                  539.61    
...
Total generated tokens:                  517120    
Total generated tokens (retokenized):    488615    
Request throughput (req/s):              0.94      
Input token throughput (tok/s):          313.74    
Output token throughput (tok/s):         958.33    
Accept length:                           1.24      
...
Mean E2E Latency (ms):                   288263.21 
Mean TTFT (ms):                          238814.62 
Mean TPOT (ms):                          48.34     
==================================================
```

### Questions / Context
1. Is the extreme TTFT overhead a known SGLang integration issue with vision tokens, and how can we fix it?
2. Given the low accept length (1.24), is our training data (12K, 2 epochs) insufficient, and what training scale do you recommend for a 32B model?

Any guidance on debugging this integration would be highly appreciated!

**Comment by jianc99** [comment]
We haven’t tried using SpecForge to train a DFlash draft model for VL models yet. However, I don’t think the issue lies with the DFlash draft model itself. Our Qwen3.5-9B-DFlash model was trained purely on text, but it still performs well on vision inputs:

<img width="1092" height="648" alt="Image" src="https://github.com/user-attachments/assets/1c2ee880-432c-4842-943c-0613e2b44c0f" />

I’ll take a closer look at the SpecForge training implementation.

**Comment by Liac-li** [comment]
Thanks for the quick response and the insightful benchmark results. I will check for any bugs in the `SpecForge` implementation.

And I’m especially excited about the potential release of the training code or scripts—it would be a huge help to the community for reproducing and building on these results.
